opacus/optimizers/adaclipoptimizer.py

```diff
@@ -0,0 +1,168 @@
+        self.sample_size = 0
+        self.unclipped_num = 0
+
+    def zero_grad(self, set_to_none: bool = False):
+        """
+        Clear gradients, self.sample_size and self.unclipped_num
+        """
+
+        if set_to_none is False:
+            logger.info(
+                "Despite set_to_none is set to False, "
+                "opacus will set p.grad_sample and p.summed_grad to None due to "
+                "non-trivial gradient accumulation behaviour"
+            )
+
+        for p in self.params:
+            p.grad_sample = None
+
+            if not self._is_last_step_skipped:
+                p.summed_grad = None
+
+        self.original_optimizer.zero_grad(set_to_none)
+
+        self.sample_size = 0
+        self.unclipped_num = 0
+
+    def clip_and_accumulate(self):
+        per_param_norms = [
+            g.view(len(g), -1).norm(2, dim=-1) for g in self.grad_samples
+        ]
+        per_sample_norms = torch.stack(per_param_norms, dim=1).norm(2, dim=1)
+        per_sample_clip_factor = (self.max_grad_norm / (per_sample_norms + 1e-6)).clamp(
+            max=1.0
+        )
+
+        self.sample_size += len(per_sample_clip_factor)
+        self.unclipped_num += (
+            len(per_sample_clip_factor) - (per_sample_clip_factor < 1).sum()
+        )
+
+        for p in self.params:
+            _check_processed_flag(p.grad_sample)
+
+            grad_sample = _get_flat_grad_sample(p)
+            grad = torch.einsum("i,i...", per_sample_clip_factor, grad_sample)
+
+            if p.summed_grad is not None:
+                p.summed_grad += grad
+            else:
+                p.summed_grad = grad
+
+            _mark_as_processed(p.grad_sample)
+
+    def add_noise(self):
+        for p in self.params:
+            _check_processed_flag(p.summed_grad)
+
+            noise = _generate_noise(
+                std=self.noise_multiplier * self.max_grad_norm,
+                reference=p.summed_grad,
+                generator=self.generator,
+                secure_mode=self.secure_mode,
+            )
+            p.grad = (p.summed_grad + noise).view_as(p.grad)
+
+            _mark_as_processed(p.summed_grad)
+
+        noise = _generate_noise(
+            std=self.fraction_std,
+            reference=self.unclipped_num,
+            generator=self.generator,
+        )
+
+        self.unclipped_num = float(self.unclipped_num)
+        self.unclipped_num += noise
+
+    def scale_grad(self):
+        """
+        Update clipping bound based on unclipped fraction
+        """
+        unclipped_frac = self.unclipped_num / self.sample_size
+        self.max_grad_norm *= torch.exp(
+            -self.clipbound_learning_rate
+            * (unclipped_frac - self.target_unclipped_quantile)
+        )
+        if self.max_grad_norm > self.max_clipbound:
+            self.max_grad_norm = self.max_clipbound
+        elif self.max_grad_norm < self.min_clipbound:
+            self.max_grad_norm = self.min_clipbound
```

Comment:
I think we can put all the code above in a function update_max_grad_norm(). While I think calling it under scale_grad works, it is not ideal as it is not the purpose of the scale_grad function.
I would recommend to add it in the pre_step, something like that:

```
def pre_step(
        self, closure: Optional[Callable[[], float]] = None
    ) -> Optional[float]:
    pre_step_full = super().pre_step()
    if pre_step_full:
        self.update_max_grad_norm()
    return pre_step_full
    
```

Reply:
Thank you so much for your suggestions! All resolved!